curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -300,315 +301,345 @@ FileCacheManager::FindOrCreateChunkCacheManager(uint64_t index) {
-            uint32_t prefetchBlocks = s3ClientAdaptor_->GetPrefetchBlocks();
-            std::vector<std::pair<std::string, uint64_t>> prefetchObjs;
-            for (int count = 0; count < prefetchBlocks; count++) {
-                std::string name = curvefs::common::s3util::GenObjName(
-                    iter->chunkId, blockIndexTmp, iter->compaction, iter->fsId,
-                    iter->inodeId);
-                uint64_t readLen = (blockIndexTmp + 1) * blockSize;
-                if (readLen > fileLen) {
-                    VLOG(6) << "end, redLen :" << readLen
-                            << ", fileLen: " << fileLen << ", blockIndexTmp "
-                            << blockIndexTmp;
-                    readLen = fileLen - blockIndexTmp * blockSize;
-                    prefetchObjs.push_back(std::make_pair(name, readLen));
-                    break;
-                } else {
-                    prefetchObjs.push_back(std::make_pair(name, blockSize));
-                }
+bool FileCacheManager::ReadKVRequestFromLocalCache(const std::string &name,
+                                                      char *databuf,
+                                                      uint64_t offset,
+                                                      uint64_t len) {
+    uint64_t start = butil::cpuwide_time_us();
 
-                blockIndexTmp++;
-                if (blockIndexTmp >= chunkSize / blockSize) {
-                    break;
-                }
-            }
-            // prefetch object from s3
-            PrefetchS3Objs(prefetchObjs);
+    bool mayCached = s3ClientAdaptor_->HasDiskCache() &&
+                     s3ClientAdaptor_->GetDiskCacheManager()->IsCached(name);
+    if (!mayCached) {
+        return false;
+    }
+
+    if (0 > s3ClientAdaptor_->GetDiskCacheManager()->Read(name, databuf, offset,
+                                                          len)) {
+        LOG(WARNING) << "object " << name << " not cached in disk";
+        return false;
+    }
+
+    if (s3ClientAdaptor_->s3Metric_.get()) {
+        s3ClientAdaptor_->CollectMetrics(
+            &s3ClientAdaptor_->s3Metric_->adaptorReadS3, len, start);
+    }
+    return true;
+}
+
+bool FileCacheManager::ReadKVRequestFromRemoteCache(const std::string &name,
+                                                    char *databuf,
+                                                    uint64_t offset,
+                                                    uint64_t length) {
+    if (!g_kvClientManager) {
+        return false;
+    }
+
+    return g_kvClientManager->Get(name, databuf, offset, length);
+}
+
+bool FileCacheManager::ReadKVRequestFromS3(const std::string &name,
+                                           char *databuf, uint64_t offset,
+                                           uint64_t length, int *ret) {
+    uint64_t start = butil::cpuwide_time_us();
+    *ret = s3ClientAdaptor_->GetS3Client()->Download(name, databuf, offset,
```

Comment:
There is only the logic of sync download here, do you not consider async download?

Reply:
Yes, it's a  remaining problem and is fixed by #2080